File: src/models/Alien.java
package models;

import abstracts.AbstractGameItem;
import constants.GameConstants;
import utils.ImageManager;

import java.awt.*;
import java.awt.geom.Point2D;
import java.util.ArrayList;

public class Alien extends AbstractGameItem implements GameConstants {
    private static final ArrayList<Alien> aliens = new ArrayList<>();
    private final boolean isStrong;
    private int livesLeft;

    public Alien(boolean isStrong) {
        super();
        this.isStrong = isStrong;
        livesLeft = isStrong ? 2 : 1;
        aliens.add(this);
    }

    public static ArrayList<Alien> getAliens() {
        return aliens;
    }

    public boolean getIsStrong() {
        return isStrong;
    }

    public int getLivesLeft() {
        return livesLeft;
    }

    public void setLivesLeft(int livesLeft) {
        this.livesLeft = livesLeft;
    }


    @Override
    public double getTravelSpeed() {
        return ALIEN_TRAVEL_SPEED;
    }

    @Override
    public void updatePosition(float deltaTime) {
        Point2D.Double pos = getPosition();
        double newY = pos.getY() + getTravelSpeed() * deltaTime;

        setPosition(new Point2D.Double(pos.getX(), newY));
    }

    @Override
    protected void draw(Graphics g) {
        if (isValid()) {
            g.drawImage(getAsset(), (int) getPosition().getX(), (int) getPosition().getY(), null);
        }
    }

    @Override
    public void setIsValid() {
        this.isValid = livesLeft > 0;
    }

    public static class RedAlien extends Alien {
        public RedAlien(boolean isStrong) {
            super(isStrong);
            setAsset(ImageManager.load(RED_ALIEN_ASSET_PATH));
        }
    }

    public static class GreenAlien extends Alien {
        public GreenAlien(boolean isStrong) {
            super(isStrong);
            setAsset(ImageManager.load(GREEN_ALIEN_ASSET_PATH));
        }
    }

    public static class YellowAlien extends Alien {
        public YellowAlien(boolean isStrong) {
            super(isStrong);
            setAsset(ImageManager.load(YELLOW_ALIEN_ASSET_PATH));
        }
    }

    public static class ExtraAlien extends Alien {
        public ExtraAlien(boolean isStrong) {
            super(isStrong);
            setAsset(ImageManager.load(EXTRA_ALIEN_ASSET_PATH));
        }
    }
}

File: src/items/Alien.kt
package items

import abstracts.AbstractGameItem
import constants.GameConstants
import engine.ImageManager.load
import java.awt.Graphics
import java.awt.geom.Point2D

open class Alien(val isStrong: Boolean, private val player: Player) : AbstractGameItem(), GameConstants {
    @JvmField
    var livesLeft: Int

    init {
        livesLeft = if (isStrong) 2 else 1 // this does not work for some reason
        aliens.add(this)
    }

    override val travelSpeed: Double
        get() = GameConstants.ALIEN_TRAVEL_SPEED

    public override fun updatePosition(deltaTime: Float) {
        val pos = position
        val newY = pos.getY() + travelSpeed * deltaTime
        position = Point2D.Double(pos.getX(), newY)

//        if (newY >= SCREEN_HEIGHT) {
//            player.setLivesLeft(player.getLivesLeft() - 1);
//            setIsValid(false);
//            System.out.printf("Yes: %s\n", this.getClass().getName());
//        }
    }

    override fun draw(g: Graphics?) {
        if (isValid()) {
            g!!.drawImage(asset, position.getX().toInt(), position.getY().toInt(), null)
        }
    }

    public override fun setIsValid() {
        isValid = livesLeft > 0
    }

    override fun setIsValid(isValid: Boolean) {
        // ignored
    }

    class RedAlien(isStrong: Boolean, player: Player) : Alien(isStrong, player) {
        init {
            asset = load(GameConstants.RED_ALIEN_ASSET_PATH)
        }
    }

    class GreenAlien(isStrong: Boolean, player: Player) : Alien(isStrong, player) {
        init {
            asset = load(GameConstants.GREEN_ALIEN_ASSET_PATH)
        }
    }

    class YellowAlien(isStrong: Boolean, player: Player) : Alien(isStrong, player) {
        init {
            asset = load(GameConstants.YELLOW_ALIEN_ASSET_PATH)
        }
    }

    class ExtraAlien(isStrong: Boolean, player: Player) : Alien(isStrong, player) {
        init {
            asset = load(GameConstants.EXTRA_ALIEN_ASSET_PATH)
        }
    }

    companion object {
        @JvmField
        val aliens = ArrayList<Alien>()
    }
}
